tools/build_paks/build_paks.py

```diff
@@ -0,0 +1,186 @@
+    txt_file   = os.path.realpath(os.path.join(output_dir, pak_name + "-" + args.version_tag + "-" + platform + ".txt"))
+    pak_file   = os.path.realpath(os.path.join(output_dir, pak_name + "-" + args.version_tag + "-" + platform + ".pak"))
+
+    # create symlinks
+    for symlink_dir_physical, symlink_dir_virtual in symlink_dirs.items():
+        if spear.path_exists(symlink_dir_virtual):
+            spear.log(f"    File or directory or symlink exists, removing: {symlink_dir_virtual}")
+            spear.remove_path(symlink_dir_virtual)
+        spear.log(f"    Creating symlink: {symlink_dir_virtual} -> {symlink_dir_physical}")
+        os.symlink(symlink_dir_physical, symlink_dir_virtual)
+
+    # now that we have created our symlinks, check that the Content/Scenes dir contains exactly expected_unreal_project_content_scene_dirs
+    if expected_unreal_project_content_scene_dirs is not None:
+        ignore_names = [".DS_Store"]
+        unreal_project_content_scene_dirs = { os.path.basename(x) for x in os.listdir(unreal_project_content_scenes_dir) if x not in ignore_names }
+        assert unreal_project_content_scene_dirs == set(expected_unreal_project_content_scene_dirs)
+
+    # cook project, see https://docs.unrealengine.com/5.2/en-US/SharingAndReleasing/Deployment/Cooking for more details
+    cmd = [
+        unreal_editor_bin,
+        uproject,
+        "-run=Cook",
+        "-targetplatform=" + platform,
+        "-unattended",                           # don't require any user input
+        "-iterate",                              # only cook content that needs to be updated
+        "-fileopenlog",                          # generate a log of which files are opened in which order
+        "-ddc=InstalledDerivedDataBackendGraph", # use the default cache location for installed (i.e., not source) builds of the engine
+        "-unversioned",                          # save all of the cooked packages without versions
+        "-stdout",                               # ensure log output is written to the terminal 
+        "-fullstdoutlogoutput",                  # ensure log output is written to the terminal
+        "-nologtimes"]                           # don't print timestamps next to log messages twice
+    spear.log(f"    Executing: {' '.join(cmd)}")
+    subprocess.run(cmd, check=True)
+
+    # create the output_dir
+    os.makedirs(output_dir, exist_ok=True)
+
+    # create manifest file in output dir
+    with open(txt_file, mode="w") as f:
+        for cooked_include_dir in cooked_include_dirs:
+            for cooked_include_file in glob.glob(os.path.realpath(os.path.join(cooked_include_dir, "**", "*.*")), recursive=True):
+                cooked_include_file_posix = cooked_include_file.replace(ntpath.sep, posixpath.sep)
+                mod_file =  [ cooked_include_file_posix[i:] for i in range(len(cooked_include_file_posix)) if cooked_include_file_posix.startswith("Engine" + posixpath.sep, i) ]
```

Comment:
This block of code is confusing. What is this code doing? Why do we need to explicitly materialize this array of all possible suffixes?

Reply:
```
    unreal_project_default_manifest_file = os.path.realpath(os.path.join(unreal_project_cooked_dir, "SpearSim", "Metadata", "packagestore.manifest"))

    # get asset files that are available by default
    with open(unreal_project_default_manifest_file) as f:
        default_manifest = json.load(f)
        default_manifest_asset_names = [ default_manifest_asset["Path"] for default_manifest_asset in default_manifest["Files"] ]

    # create manifest file in output dir
    with open(txt_file, mode="w") as f:
        for cooked_include_dir in cooked_include_dirs:
            for cooked_include_file in glob.glob(os.path.realpath(os.path.join(cooked_include_dir, "**", "*.*")), recursive=True):
                assert cooked_include_file.startswith(unreal_project_cooked_dir)
                asset_name = cooked_include_file.removeprefix(os.path.join(unreal_project_cooked_dir, os.path.sep)).replace(ntpath.sep, posixpath.sep)
                if asset_name not in default_manifest_asset_names:
                    cooked_include_file_posix = cooked_include_file.replace(ntpath.sep, posixpath.sep)
                    assert cooked_include_file_posix.startswith(unreal_project_cooked_dir_posix)
                    cooked_mount_file_posix = posixpath.join("..", "..", "..", cooked_include_file_posix.replace(unreal_project_cooked_dir_posix + posixpath.sep, ""))
                    f.write(f'"{cooked_include_file_posix}" "{cooked_mount_file_posix}" "" \n')
```

Reply:
Aah, forgot about the `removeprefix` functionality. It's good that we also recently upgraded to python 3.9 as `removeprefix` was introducted in pythobn 3.9.

Reply:
In Python 3.8 you could also do `assert str.startswith(bad_str); str.replace(bad_str, "", 1)` to indicate that you only want to replace remove the 1st occurance of `bad_str`. But I agree that `removeprefix` is cleaner.